src/api/api_context.cpp

```diff
@@ -38,9 +38,10 @@ namespace api {
     void object::inc_ref() { ++m_ref_count; }
 
     void object::dec_ref() { SASSERT(m_ref_count > 0); if (--m_ref_count == 0) m_context.del_object(this); }
-    
+
     unsigned context::add_object(api::object* o) {
         flush_objects();
+        lock_guard lock(m_mux);
```

Comment:
what is the race that caused adding a lock guard here?
add_object on a context should only be called within the same thread.

Reply:
`flush_objects` can be called concurrently by the GC, which also modifies `m_allocated_objects`. It's the same reason for having locks in .NET/Java and DecRefQueues to help with contention.

I believe that in OCaml 5 we could possibly get away with it, if we remove all the `*_dec_ref` calls in `z3.ml`, because the OCaml 5 GC doesn't parallelize internally, but I'm sure someone's already working on that...

Actually, looking at it again, I think I may have tricked myself with the call to `flush_objects` in `Z3_solver_dec_ref`. I'll try it without.

Reply:
Yeah, I think it's fine without those changes, although I will have to get rid of the `*_dec_ref`s in `z3.ml`. I do see a significant increase in non-deterministic unknowns, but that's probably a side-effect.